enumerables.rb

```diff
@@ -0,0 +1,273 @@
+module Enumerable
+  def my_each(&block)
+    return to_enum unless block_given?
+
+    i = 0
+    if is_a? Range
+      iterable_array = self.to_a
+    elsif is_a? Array
+      iterable_array = self
+    elsif is_a? Hash
+      iterable_array = self.to_a
+      if block.arity == 0
+        yield
+      elsif block.arity == 1
+        i = 0
+        while i < iterable_array.length
+          yield(iterable_array[i])
+          i += 1
+        end
+        return self
+      elsif block.arity >= 2
+        i = 0
+        while i < iterable_array.length
+          yield(iterable_array[i][0], iterable_array[i][1])
+          i += 1
+        end
+      end
+      self
+    end
+
+    while i <= iterable_array.length - 1
+      yield(iterable_array[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
```

Comment:
## #my_each_with_index:

- [ ] allow #my_each_with_index to work with ranges: `undefined local variable or method `length' for 5..50:Range`